File: LeetCode/List&LinkList/876.middle-of-the-linked-list.py
#
# @lc app=leetcode id=876 lang=python3
#
# [876] Middle of the Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        p_fast = head
        p_slow = head
        while(p_fast and p_fast.next):
            p_slow = p_slow.next
            p_fast = p_fast.next.next
        return p_slow
        
# 快慢指针慢指针走一，快指针走二，快指针走到的时候，慢指针真指向就是终点
# @lc code=end

File: LeetCode_Go/List&LinkedList/876.Middle-of-the-Linked-List.go
/*
 * @lc app=leetcode id=908 lang=golang
 *
 * [876] Middle of the Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// Given a linked list, return a middle node of linked list.

// fast and slow pointers. both start from the head. fast pointer 2* speed. when fast pointer reaches the end of the linked list, slow pointer reaches the middle of the linked list
func middleNode(head *ListNode) *ListNode {
	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

// @lc code=end

